mlir/include/mlir/Dialect/EmitC/IR/EmitC.td

```diff
@@ -402,4 +403,102 @@ def EmitC_VariableOp : EmitC_Op<"variable", []> {
   let hasVerifier = 1;
 }
 
+def EmitC_AssignOp : EmitC_Op<"assign", [MemoryEffects<[MemWrite]>]> {
```

Comment:
I think we can be a bit more strict with the MemoryEffects, i.e. `var` is written and `value` is read. This kind of resembles the [memref.copy](https://github.com/llvm/llvm-project/blob/7675f541f75baa20e8ec007cd625a837e89fc01f/mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td#L518) op.

And I guess the Variable op could get the [MemAlloc effect](https://github.com/llvm/llvm-project/blob/7675f541f75baa20e8ec007cd625a837e89fc01f/mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td#L75).

Reply:
W.R.T the underlying C variables to be emitted, the assign op takes an SSA Value as the value to assign, not necessarily another variable. The emitter may or may not generate a C variable for that value, e.g. emitc::literal, so a memread may or may not happen, which is consistent with all other ops taking SSA Values as operands.

So on second thought adding memory effects to the assign op at this point may be a bit premature, since as you point out allocation/reading/writing are currently not modeled explicitly in the dialect at all. Being an output dialect such semantics may be redundant, but that probably deserves a separate discussion.
Suggest to remove the MemoryEffects from this patch, leaving assign's memory semantics to be handled as part of the whole dialect. What do you think?

Reply:
Ah yes, I didn't think this through. Sounds good to me to remove it for now and revisit later if needed.